sdk/core/core-util/src/reshape.ts

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+interface SelectorPropertyName {
+  type: "propertyName";
+  value: string;
+}
+
+interface SelectorArrayIterator {
+  type: "arrayIterator";
+  remainingSelector: string;
+}
+
+type SelectorPart = SelectorArrayIterator | SelectorPropertyName;
+
+function* selectorParts(selector: string): Iterable<SelectorPart> {
```

Comment:
not sure if it's possible, but could we leverage jsonpath to do the selection ? I feel like that's more generally used ?

Reply:
If we are considering jsonpath to do the selection, could it be used in the customization? like in the original autorest modelerfour, they use directive for customized transformation, and the selector is also using jsonpath.

Reply:
May I ask a silly question here,  I am not sure I get the point of why jsonpath being made into an RFC matters to us ?  😊

My concerns of supporting selector by ourself is that, it could possibly be buggy as I am not sure if we have all cases considered ? I am also not sure if it's a good idea that service team who wants to write their own customization serialize and deserialize function and would have to learn the grammer we invented here?

Reply:
Personally I would prefer to have jsonpath to locate the objects with less learning curve. Is it possible our selectors are just selected subsets of jsonpath?

Reply:
I suppose there are two different considerations here, one is taking a dependency on some third party jsonpath library, the other is supporting the full jsonpath syntax. 

The former I'm not that interested in, since it feels unnecessary to take a dependency for the problem we have (provide an efficient and debuggable primitive for renaming properties and mapping values) and we wouldn't be able to easily optimize any code in the dependency itself.

As to the latter, if we're saying "hey this is a syntax people know, so we should use it", that's great except since it's not a standard, there could be different implementations with different behaviors/syntax, so it's not really buying us as much familiarity with an existing syntax. In addition, since this isn't a consumer-facing function, I'm not sure I buy the learning curve being difficult or there being a high need for familiarity.

To address @MaryGao's point, if we wanted to align the syntax we totally could, but it would simply change something like `a.b.c[]d` into `$.a.b.c[*].d` and all we've really done is:

- add a questionably useful `$.` prefix to every single selector
- forced you to write `*` in between the array braces
- put an extra `.` after the brace syntax to access properties inside an array element. 

I'm not convinced that adding these extra characters greatly improves readability. 

I'm curious about this statement:

>  service team who wants to write their own customization serialize and deserialize function 

Is this a scenario we support today? And if so, why would they necessarily have to use this helper? Presumably if they are going through the work of writing a custom TS function that converts one shape to another, they could do so efficiently without any helpers, much like we did in past track 2 dataplane packages.

Reply:
Wonder a case like `{ a: {b: {c: 1}}, b: {c: 2} }`, if we pass `b.c`, would it be confusing which part we are referring to ? 

> Is this a scenario we support today? And if so, why would they necessarily have to use this helper?

Can't argue with that, customer can definitely write customization helper as they like.

Reply:
> Wonder a case like { a: {b: {c: 1}}, b: {c: 2} }, if we pass b.c, would it be confusing which part we are referring to ?

I mean if a model from a service has multiple sections with repeated names it could be a little confusing to look at, but is it more confusing than reading `parsedBody.a.b.c` vs `parsedBody.b.c` ?